rclcpp/include/rclcpp/experimental/subscription_intra_process.hpp

```diff
@@ -115,13 +116,34 @@ class SubscriptionIntraProcess : public SubscriptionIntraProcessBase
   bool
   is_ready(rcl_wait_set_t * wait_set)
   {
-    (void)wait_set;
+    (void) wait_set;
     return buffer_->has_data();
   }
 
-  void execute()
+  void take_data(std::shared_ptr<void> & data)
   {
-    execute_impl<CallbackMessageT>();
+    if (data) {
+      throw std::runtime_error("'data' is not empty");
+    }
+
+    ConstMessageSharedPtr shared_msg;
+    MessageUniquePtr unique_msg;
+
+    if (any_callback_.use_take_shared_method()) {
+      shared_msg = buffer_->consume_shared();
+    } else {
+      unique_msg = buffer_->consume_unique();
+    }
+    data = std::static_pointer_cast<void>(
+      std::make_shared<std::pair<ConstMessageSharedPtr, MessageUniquePtr>>(
+        std::pair<ConstMessageSharedPtr, MessageUniquePtr>(
+          shared_msg, std::move(unique_msg)))
+    );
+  }
+
+  void execute(std::shared_ptr<void> & data)
```

Comment:
Given that we are changing public API, do we think we should maintain a deprecated version of this API without the data?

Reply:
The issue is that we have to still call only one from the executor.
Maybe we can always call `execute(data)` from the executor, and provide the following defaults:

```
class Waitable {
  virtual std::shared_ptr<void> take_data() {return nullptr};
  virtual execute(std::shared_ptr<void> data) {(void)data; this->execute()};
  virtual execute() = 0;
}
```

So you can choose to either override only `execute()`, or `execute(data)` and `take_data()`.
We could even make use of that in the places were `take_data` returns `nullptr`.

@audrow @clalancette does that sound good to you?

Reply:
I realized that I cannot really use:
```cpp
  virtual execute() = 0;
```

but rather need:
```cpp
  virtual execute() {throw std::runtime_error{"either execute() or execute(std::shared_ptr<void> &) need to be overriden"};};
```

Because you cannot have a pure virtual method.

I think I would rather break API and add a comment in the release notes than do the above.
I don't expect much people creating custom waitables out of `rclcpp`.

Reply:
All right, that's OK with me.  I will suggest a release note for Galactic, then, just in case people are creating custom waitables.

Reply:
👍